### DISCUSSION TITLE
How to manually overwrite the link ratios?

### DISCUSSION THREAD

**Discussion by contenrico** [casact/chainladder-python#443] [Q&A]
Hello,

I'm trying to manually overwrite the link ratios calculated through `triangle.link_ratio`.

As `triangle.link_ratio` is also a Triangle object, I'm guessing I can use `.loc[]` to substitute specific link ratios with my manual inputs. However, I can't find a way to maintain these all the down the pipeline.

For example, when I subsequently use `cl.Development().fit(triangle)`, it goes back to using the original link ratios. However, I'd like to be able to use all the functionalities of `cl.Development` and `cl.Chainladder()` with my modified link ratios. Is it possible to do that?

Thanks!

**Comment by jbogaardt** [comment]
Hi @contenrico. Neither the `Development` estimator nor the `Triangle` supports modifying the link ratios directly. The `Triangle` assumes link ratios to be a function of the underlying data and `Development` estimator allows for various options of selecting a pattern from link ratios, but not in a way that modifies them.

To achieve the functionality you're looking for, I could suggest doing the modification outside of an estimator and using `DevelopmentConstant` to get it back into a chainladder-compliant format.

For example:

```python
import chainladder as cl
import numpy as np

raa = cl.load_sample('raa')

# Modify link ratios however you want. Here I am naively limiting the low and high of my link ratios directly 
limited_link_ratios = np.maximum(np.minimum(raa.link_ratio, 5), 1.0)

# Convert link ratios into a selected pattern stored as a dictionary of (age,  ldf) pairs.
selected_patterns = dict(
    zip(raa.development[:-1], 
        limited_link_ratios.mean().to_frame().iloc[0])
)

# Use DevelopmentConstant to leverage your custom patterns in a compliant way
pipe = cl.Pipeline([
    ('dev', cl.DevelopmentConstant(patterns=selected_patterns, style='ldf')),
    ('model', cl.Chainladder())]
)

# Fit a chainladder model to your selected patterns and explore its properties.
pipe.fit(raa).named_steps.model.ibnr_
```

**Reply by contenrico** [reply]
Hi @jbogaardt 

Many thanks for your reply.

This approach could work, however I was wondering if there an easy way to have more control over how the LDFs are calculated? I'm referring to this snippet:

`limited_link_ratios.mean().to_frame().iloc[0]`

This is just calculating the simple mean for all origins - is there a way to control the type of average (i.e. select the volume average option) and the number of periods considered? Or would I have to create a custom function for that?

Thanks!

**Comment by jbogaardt** [comment]
The `Development` estimator is a pretty flexible estimator for selecting various development patterns. However, it does NOT support modifying the age-to-age factors of the underlying Triangle itself. I probably don't I understand your use case fully. Does the `Development` estimator not provide enough flexibility in selecting patterns? Or can you describe the use case that requires modifying the underlying link ratios directly before selecting patterns?

**Reply by contenrico** [reply]
My specific use case would involve Nth-root smoothing of the age-to-age factors.

What I would like to do, for example, is take three LDFs in an origin year (let's say 1.5, 3, 1.25) and replace them with the cube root of (1.5 * 3 * 1.25). In this case, I would replace 1.5, 3, 1.25 with 1.78, 1.78, 1.78.

I don't think the `Development` estimator can do that, so I was trying to find a way to replace link ratios directly, however I would like to maintain the functionalities of  the `Development` estimator (such as the volume average). Is there potentially another way to achieve the same type of smoothing?

**Reply by jbogaardt** [reply]
I see. This is geometric averaging which has been requested in #272. I think this is a worthwhile enhancement. Still, it doesn't do you much good until implemented. However, similar to my example above, you can probably get pretty creative with what currently exists along the lines I mention above. In the case of a 3-year geometric average, you could do this:
```python
import chainladder as cl
import numpy as np

raa = cl.load_sample('raa')
n = 3

# Modify link ratios however you want. Here is he cubed-root 3 year most recent average
link_ratios = cl.Development(n_periods=n).fit_transform(raa).link_ratio.prod(axis=2) ** (1/n)


# Convert link ratios into a selected pattern stored as a dictionary of (age,  ldf) pairs.
selected_patterns = dict(
    zip(raa.development[:-1], link_ratios.to_frame().iloc[0])
)

# Use DevelopmentConstant to leverage your custom patterns in a compliant way
pipe = cl.Pipeline([
    ('dev', cl.DevelopmentConstant(patterns=selected_patterns, style='ldf')),
    ('model', cl.Chainladder())]
)

# Fit a chainladder model to your selected patterns and explore its properties.
pipe.fit(raa).named_steps.model.ibnr_
```

**Reply by contenrico** [reply]
Thanks @jbogaardt.

Just for clarity, my request is not exactly the same as #272. My understanding is that this request is asking for a geometric mean across the columns, whereas my request is about a geometric mean across a row (which results in the smoothing of a pattern for a given origin year).

So starting from a triangle like this:
![image](https://github.com/casact/chainladder-python/assets/137546866/a88c1659-9a50-4f53-a10c-39ee490d4164)

I would like to be able to replace sets of factors with their geometric mean like so (red boxes):
![image](https://github.com/casact/chainladder-python/assets/137546866/4c257a3e-9306-4d6e-ac44-000fe9d3e29c)
